scio-core/src/main/scala/com/spotify/scio/values/BloomFilter.scala

```diff
@@ -0,0 +1,202 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.scio.values
+
+import java.io._
+
+import com.google.common.hash.{Funnel, BloomFilter => gBloomFilter}
+import com.spotify.scio.annotations.experimental
+import com.spotify.scio.coders.Coder
+
+/**
+ * Bloom Filter - a probabilistic data structure to test approximate presence of an element.
+ *
+ * Operations
+ * 1) create: Hash the value k times, updating the bitfield at
+ *            the index equal to each hashed value
+ * 2) query: hash the value k times.  If there are k collisions, then return true; otherwise false.
+ *
+ * http://en.wikipedia.org/wiki/Bloom_filter
+ *
+ * Implemented as an immutable wrapper over Guava's Bloom Filter.
+ */
+@SerialVersionUID(1L)
+final case class BloomFilter[T] private (
+  private val internal: gBloomFilter[T],
+  private val funnel: Funnel[T]
+) extends ApproxFilter[T] {
+
+  /**
+   * Returns the probability that [[mayBeContains(t: T)]] will erroneously return `true`
+   * for an element that was not actually present the colleciton from which this [[BloomFilter]]
+   * was built.
+   *
+   * Ideally, this number should be close to the `fpProb` parameter passed to the [[BloomFilter#apply]]
+   * when this filter was built or smaller. If it is significantly higher, it is usually the
+   * case that too many elements (more than expected) were present in the original collection
+   * from which this [[BloomFilter]] was built.
+   */
+  val expectedFpp: Double = internal.expectedFpp()
+
+  /**
+   * Returns an estimate for the total number of distinct elements that have been added to this
+   * [[BloomFilter]]. This approximation is reasonably accurate if it does not exceed the value of
+   * `expectedInsertions` that was used when constructing the filter.
+   */
+  val approximateElementCount: Long = internal.approximateElementCount()
+
+  /**
+   * Check approximate presence of an element in the BloomFilter
+   *
+   * @return true if the element may be present, false if it is definitely not present.
+   */
+  def mayBeContains(t: T): Boolean = internal.mightContain(t)
+
+  // Java Serialization
+  private def writeObject(out: ObjectOutputStream): Unit = {
+    out.writeObject(funnel)
+    internal.writeTo(out)
+  }
+
+  private def readObject(in: ObjectInputStream): Unit = {
+    val funnel = in.readObject().asInstanceOf[Funnel[T]]
+    val internal = gBloomFilter.readFrom(in, funnel)
+    BloomFilter.setField("funnel", funnel)
+    BloomFilter.setField("internal", internal)
+  }
+
+  /**
+   * Add an element to this [[BloomFilter]]. It creates a copy of the underlying
+   * structure.
+   *
+   * For creating BloomFilters from large collections, use [[BloomFilter#apply(Iterable)]]
+   * instead.
+   */
+  @experimental
+  def put(t: T): BloomFilter[T] = {
```

Comment:
Do we really want to expose these 2? They will be misused and sort of defeats the purpose of wrapping mutable Guava BF. I'd at least make them package private and rename to something that indicates copy, like `+/++`.

Reply:
I am not sure if we should keep this. The use case in Scio would be to add more elements at different stages.